keanu-python/keanu/vertex.py

```diff
@@ -1,30 +1,115 @@
 import numpy as np
 import numbers
 
+import keanu as kn
 from keanu.context import KeanuContext
 from keanu.base import JavaObjectWrapper
-from keanu.const import Const
 
 
 context = KeanuContext()
 
 
 class VertexOps:
-    def __gt__(self, other):
-        from keanu.generated.vertex import GreaterThan
-        return GreaterThan(self, other)
+    """
+    __array_ufunc__ is a NumPy thing that enables you to intercept and handle the numpy operation.
+    Without this the right operators would fail.
+    See https://docs.scipy.org/doc/numpy-1.13.0/neps/ufunc-overrides.html
+    """
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+        methods = {
+            "equal" : VertexOps.__eq__,
+            "not_equal" : VertexOps.__ne__,
+            "add" : VertexOps.__radd__,
+            "subtract" : VertexOps.__rsub__,
+            "multiply" : VertexOps.__rmul__,
+            "power" : VertexOps.__rpow__,
+            "true_divide" : VertexOps.__rtruediv__,
+            "greater" : VertexOps.__lt__,
+            "greater_equal" : VertexOps.__le__,
+            "less" : VertexOps.__gt__,
+            "less_equal" : VertexOps.__ge__,
+        }
+        if method == "__call__":
+            try:
+                dispatch_method = methods[ufunc.__name__]
+                return dispatch_method(inputs[1], inputs[0])
+            except KeyError:
+                raise NotImplementedError("NumPy ufunc of type %s not implemented" % ufunc.__name__)
+        else:
+            raise NotImplementedError("NumPy ufunc method %s not implemented" % method)
 
     def __add__(self, other):
-        from keanu.generated.vertex import Addition
-        return Addition(self, other)
+        return kn.generated.vertex.Addition(self, other)
+
+    def __radd__(self, other):
```

Comment:
not really, because `self` has a special meaning.

It would be possible for commutative operations like `+` - i.e. here I could do:

    def __radd__(self, other):
        return self.__add__(other)

but that would not work for non-commutative operations like `-`. I could do:

    def __rsub__(self, other):
        return Vertex.__sub__(other, self)

but that seems like an abuse of the concept of `self`. Overall I think it all looks better as it is.